### PR TITLE
calibrated timestamps sample

### PR DESCRIPTION
**Pull request by gpx1000**
## Description

Add a sample showing how to use Calibrated Timestamps in an existing sample.

Fixes #<issue number>

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+ */
+
+#pragma once
+
+#include "api_vulkan_sample.h"
+
+class CalibratedTimestamps : public ApiVulkanSample
+{
+  private:
+	bool                         isTimeDomainUpdated = false;        // this is just to tell if time domain update has a VK_SUCCESS in the end
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: I think, the naming convention for member variables is camel_case here. That is, isTimeDomainUpdated should probably be named is_time_domain_updated.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	bool                         isTimestampUpdated  = false;        // this is just to tell the GUI whether timestamps operation has a VK_SUCCESS in the end
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: isTimestampUpdated -> is_timestamp_updated

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	uint32_t                     time_domain_count   = 0;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: Why do you need `time_domain_count`? It's just the `time_domains.size()`.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	struct
+	{
+		VkDescriptorSetLayout models       = VK_NULL_HANDLE;
+		VkDescriptorSetLayout composition  = VK_NULL_HANDLE;
+		VkDescriptorSetLayout bloom_filter = VK_NULL_HANDLE;
+	} descriptor_set_layouts{};
+
+	struct FrameBufferAttachment
+	{
+		VkImage        image{};
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: Why do you initialize some handles with `= VK_NULL_HANDLE`, and others with `{}`?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+		VkFramebuffer         framebuffer{};
+		FrameBufferAttachment color[1]{};
+		VkRenderPass          render_pass{};
+		VkSampler             sampler{};
+	} filter_pass{};
+
+	std::vector<std::string> object_names{};
+
+  private:
+	// MODs:
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: `MODs:` ?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	static std::string read_time_domain(VkTimeDomainEXT inputTimeDomain);        // this returns a human-readable info for what time domain corresponds to what
+	void               update_time_domains();                                    // this extracts total number of time domain the (physical device has, and then sync the time domain EXT data to its vector
+	void               update_timestamps();                                      // this creates local timestamps information vector, update timestamps vector and deviation vector
+	void               init_timeDomains_and_timestamps();                        // this initializes all time domain and timestamps related variables, vectors, and booleans
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: `init_timeDomains_and_timestamps` -> `init_time_domains_and_timestamps`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	void               get_optimal_time_domain();                                // this gets the optimal time domain which has the minimal value on its max deviation.
+	void               tic(const std::string &input_tag = "Untagged");           // this marks the timestamp begin and partially updates the delta_timestamps
+	void               toc(const std::string &input_tag = "Untagged");           // this marks the timestamp ends and updates the delta_timestamps
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: I like the naming of `tic()` and `toc()`. But maybe something like `timestamp_begin()` and `timestamp_end()` is more descriptive?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+	std::array<VkClearValue, 2> clear_values_0{};
+	clear_values_0[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+	clear_values_0[1].depthStencil = {0.0f, 0};
+
+	VkRenderPassBeginInfo render_pass_begin_info_0 = vkb::initializers::render_pass_begin_info();
+	render_pass_begin_info_0.renderPass            = render_pass;
+	render_pass_begin_info_0.renderArea.offset.x   = 0;
+	render_pass_begin_info_0.renderArea.offset.y   = 0;
+	render_pass_begin_info_0.clearValueCount       = 2;
+	render_pass_begin_info_0.pClearValues          = clear_values_0.data();
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: `render_pass_begin_info_0` and `clear_values_0` are never used.
Can be removed.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 9]
```diff
@@ -0,0 +1,969 @@
...
+
+	toc("Creating Command Buffer");
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: This section does not measure the creation of a command buffer but just the initialization of some structs.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 10]
```diff
@@ -0,0 +1,969 @@
...
+
+	tic("Draw Command Buffer");
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: This section does not measure the drawing of a command buffer, but the recording/building of some command buffers.
You should either change the tags for both measurements, or do just one measurement over the complete `build_command_buffers` function, maybe tagging as "Build Command Buffers"?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 11]
```diff
@@ -0,0 +1,969 @@
...
+			clear_values_1[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_values_1[1].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_values_1[2].depthStencil = {0.0f, 0};
+
+			VkRenderPassBeginInfo render_pass_begin_info_1    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info_1.renderPass               = offscreen.render_pass;
+			render_pass_begin_info_1.framebuffer              = offscreen.framebuffer;
+			render_pass_begin_info_1.renderArea.extent.width  = offscreen.width;
+			render_pass_begin_info_1.renderArea.extent.height = offscreen.height;
+			render_pass_begin_info_1.clearValueCount          = 3;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Maybe better than plain `3`: `static_cast<uint32_t>(clear_values_1.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 12]
```diff
@@ -0,0 +1,969 @@
...
+		if (bloom)
+		{
+			std::array<VkClearValue, 2> clear_values{};
+			clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_values[1].depthStencil = {0.0f, 0};
+
+			VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info.framebuffer              = filter_pass.framebuffer;
+			render_pass_begin_info.renderPass               = filter_pass.render_pass;
+			render_pass_begin_info.clearValueCount          = 1;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Why do you specify two `clear_values`, but use just one?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 13]
```diff
@@ -0,0 +1,969 @@
...
+
+		{
+			std::array<VkClearValue, 2> clear_values{};
+			clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_values[1].depthStencil = {0.0f, 0};
+
+			VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info.framebuffer              = framebuffers[i];
+			render_pass_begin_info.renderPass               = render_pass;
+			render_pass_begin_info.clearValueCount          = 2;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Maybe better than plain `2`, again: `static_cast<uint32_t>(clear_values.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 14]
```diff
@@ -0,0 +1,969 @@
...
+	image_view_create_info.subresourceRange.baseMipLevel   = 0;
+	image_view_create_info.subresourceRange.levelCount     = 1;
+	image_view_create_info.subresourceRange.baseArrayLayer = 0;
+	image_view_create_info.subresourceRange.layerCount     = 1;
+	image_view_create_info.image                           = attachment->image;
+	VK_CHECK(vkCreateImageView(get_device().get_handle(), &image_view_create_info, nullptr, &attachment->view));
+}
+
+void CalibratedTimestamps::prepare_offscreen_buffer()
+{
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:246]: The current version of the HDR sample uses some algorithm to determine the color format. You should use that here as well.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 15]
```diff
@@ -0,0 +1,969 @@
...
+		dependencies[1].dstAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
+		dependencies[1].dependencyFlags = VK_DEPENDENCY_BY_REGION_BIT;
+
+		VkRenderPassCreateInfo render_pass_create_info = {};
+		render_pass_create_info.sType                  = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
+		render_pass_create_info.pAttachments           = attachment_descriptions.data();
+		render_pass_create_info.attachmentCount        = static_cast<uint32_t>(attachment_descriptions.size());
+		render_pass_create_info.subpassCount           = 1;
+		render_pass_create_info.pSubpasses             = &subpass;
+		render_pass_create_info.dependencyCount        = 2;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Maybe better than plain `2`: `static_cast<uint32_t>(dependencies.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 16]
```diff
@@ -0,0 +1,969 @@
...
+	image_view_create_info.subresourceRange.layerCount     = 1;
+	image_view_create_info.image                           = attachment->image;
+	VK_CHECK(vkCreateImageView(get_device().get_handle(), &image_view_create_info, nullptr, &attachment->view));
+}
+
+void CalibratedTimestamps::prepare_offscreen_buffer()
+{
+	{
+		offscreen.width  = static_cast<int>(width);
+		offscreen.height = static_cast<int>(height);
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: You should cast to `int32_t`, not to `int`.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 17]
```diff
@@ -0,0 +1,969 @@
...
+		sampler.maxAnisotropy       = 1.0f;
+		sampler.minLod              = 0.0f;
+		sampler.maxLod              = 1.0f;
+		sampler.borderColor         = VK_BORDER_COLOR_FLOAT_OPAQUE_WHITE;
+		VK_CHECK(vkCreateSampler(get_device().get_handle(), &sampler, nullptr, &offscreen.sampler));
+	}
+
+	{
+		filter_pass.width  = static_cast<int>(width);
+		filter_pass.height = static_cast<int>(height);
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: You should cast to `int32_t`, not to `int`.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 18]
```diff
@@ -0,0 +1,969 @@
...
+
+	VkPipelineVertexInputStateCreateInfo empty_input_state = vkb::initializers::pipeline_vertex_input_state_create_info();
+	pipeline_create_info.pVertexInputState                 = &empty_input_state;
+
+	shader_stages[0]                  = load_shader("hdr/composition.vert", VK_SHADER_STAGE_VERTEX_BIT);
+	shader_stages[1]                  = load_shader("hdr/composition.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+	pipeline_create_info.layout       = pipeline_layouts.composition;
+	pipeline_create_info.renderPass   = render_pass;
+	rasterization_state.cullMode      = VK_CULL_MODE_FRONT_BIT;
+	color_blend_state.attachmentCount = 1;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:638]: `1` -> `static_cast<uint32_t>(blend_attachment_states.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 19]
```diff
@@ -0,0 +1,969 @@
...
+void CalibratedTimestamps::render(float delta_time)
+{
+	if (!prepared)
+		return;
+	draw();
+	if (camera.updated)
+		update_uniform_buffers();
+}
+
+std::string CalibratedTimestamps::read_time_domain(VkTimeDomainEXT inputTimeDomain)
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: This function does not read anything, but gets you a string. I would call it `to_string`.
Morevover, it's just a local helper function, that does not need to be a member function of `CalibratedTimestamps`.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 20]
```diff
@@ -0,0 +1,969 @@
...
+		case VK_TIME_DOMAIN_CLOCK_MONOTONIC_RAW_EXT:
+			return "clock monotonic raw time domain";
+		case VK_TIME_DOMAIN_QUERY_PERFORMANCE_COUNTER_EXT:
+			return "query performance time domain";
+		default:
+			return "unknown time domain";
+	}
+}
+
+void CalibratedTimestamps::update_time_domains()
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: This function just gets you the supported time domains. Maybe just name it `get_time_domains`?
Moreover, are those time domains supposed to vary over time? Why do you call it on every `tic` and `toc`?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 21]
```diff
@@ -0,0 +1,969 @@
...
+		// Resize time domains vector:
+		time_domains.resize(static_cast<int>(time_domain_count));        // this needs static cast to int
+		// Update time_domain vector:
+		result = vkGetPhysicalDeviceCalibrateableTimeDomainsEXT(get_device().get_gpu().get_handle(), &time_domain_count, time_domains.data());
+	}
+	// Time domain is successfully updated:
+	isTimeDomainUpdated = (result == VK_SUCCESS);
+}
+
+void CalibratedTimestamps::update_timestamps()
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: I would call this function `get_timestamps`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 22]
```diff
@@ -0,0 +1,969 @@
...
+
+void CalibratedTimestamps::get_optimal_time_domain()
+{
+	init_timeDomains_and_timestamps();
+	update_time_domains();
+	update_timestamps();
+
+	if (isTimeDomainUpdated && isTimestampUpdated && (time_domain_count > 1))
+	{
+		uint64_t optimal_maxDeviation = *std::min_element(max_deviations.begin(), max_deviations.end());
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: std::min_element gets you the iterator. You don't need to find it again

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 23]
```diff
@@ -0,0 +1,969 @@
...
+
+		auto iterator_optimal = std::find(max_deviations.begin(), max_deviations.end(), optimal_maxDeviation);
+
+		if (iterator_optimal != max_deviations.end())
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: ... and, as long as there's at least one elment, the iterator will never be the `end()`. No need to check that.

**Review by asuessenbach — COMMENTED** [review]
Although I'm not a native speaker, I think the readme.md needs some substantial word-smithing.

Besides that, my understanding of the time domains is a bit different than yours. I think, there's nothing like an "optimal" time domain. There is one device time domain and one (Windows) or two (Linux) host device domains. They are not better or worse, they are different domains.
For the device time domain, you need `VkQueueFamilyProperties::timestampValidBits` and `VkPhysicalDeviceLimits::timestampPeriod` to determine nanoseconds.
While the two linux time domains already are in nanoseconds, you need the result from the windows function `QueryPerformanceFrequency` for the windows time domain to get from ticks to seconds.
Finally, the maxDeviation values could be used, somehow, to asses the quality of the actual time stamps.

**Comment by SaschaWillems** [comment]
This crashes for me at runtime in the `timestamps_begin` function in the following line:

image

The `timestamps.clear()` there looks wrong, as that vector is then never resized and always has zero elements

Same for `timestamps_end`. The vector is cleared here too and never resized before it's accessed.

**Review by SaschaWillems — CHANGES_REQUESTED** [review]
See my comment, the sample doesn't work for me at all.

**Comment by JeremyG2101** [comment]
> This crashes for me at runtime in the `timestamps_begin` function in the following line:
> 
> image
> 
> The `timestamps.clear()` there looks wrong, as that vector is then never resized and always has zero elements
> 
> Same for `timestamps_end`. The vector is cleared here too and never resized before it's accessed.

Ohh! I forgot to recall that timestamps vector must  be resized and then pass its reference/data to vkGetCalibratedTimestampsEXT.. And for some reason it doesn't crush from my end...
But you are right! Good catch and thanks!

**Comment by gpx1000** [comment]
This PR is ready for review.

**Review by SaschaWillems — DISMISSED** [review]

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 24]
```diff
@@ -0,0 +1,988 @@
...
+	is_time_domain_updated = ((result == VK_SUCCESS) && (time_domain_count > 0));
+}
+
+void CalibratedTimestamps::get_timestamps()
+{
+	// Ensures that time domain exists
+	if (is_time_domain_updated)
+	{
+		// Get calibrated timestamps:
+		VkResult result = vkGetCalibratedTimestampsEXT(get_device().get_handle(), time_domains.size(), timestamps_info.data(), timestamps.data(), max_deviations.data());
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Please `static_cast<uint32_t>(time_domains.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 25]
```diff
@@ -0,0 +1,988 @@
...
+			clear_values_1[1].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_values_1[2].depthStencil = {0.0f, 0};
+
+			VkRenderPassBeginInfo render_pass_begin_info_1    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info_1.renderPass               = offscreen.render_pass;
+			render_pass_begin_info_1.framebuffer              = offscreen.framebuffer;
+			render_pass_begin_info_1.renderArea.extent.width  = offscreen.width;
+			render_pass_begin_info_1.renderArea.extent.height = offscreen.height;
+			render_pass_begin_info_1.clearValueCount          = static_cast<uint32_t>(clear_values_1.size());
+			render_pass_begin_info_1.pClearValues             = clear_values_1.data();
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Maybe move clear_values_1 and render_pass_begin_info_1 out of the loop.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 26]
```diff
@@ -0,0 +1,988 @@
...
+			draw_model(models.objects[models.object_index], draw_cmd_buffers[i]);
+
+			vkCmdEndRenderPass(draw_cmd_buffers[i]);
+		}
+
+		if (bloom)
+		{
+			VkClearValue clear_value{};
+			clear_value.color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_value.depthStencil = {0.0f, 0};
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: `VkClearValue` is a union; setting `depthStencil` overwrites the `color`.
You should probably stay with the `color`.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 27]
```diff
@@ -0,0 +1,988 @@
...
+		color_references.push_back({1, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL});
+
+		VkAttachmentReference depth_reference = {};
+		depth_reference.attachment            = 2;
+		depth_reference.layout                = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+
+		VkSubpassDescription subpass    = {};
+		subpass.pipelineBindPoint       = VK_PIPELINE_BIND_POINT_GRAPHICS;
+		subpass.pColorAttachments       = color_references.data();
+		subpass.colorAttachmentCount    = 2;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: `2` -> `static_cast<uint32_t>(color_references.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 28]
```diff
@@ -0,0 +1,988 @@
...
+		attachment_description.finalLayout    = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+		attachment_description.format         = filter_pass.color[0].format;
+
+		std::vector<VkAttachmentReference> color_references;
+		color_references.push_back({0, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL});
+
+		VkSubpassDescription subpass = {};
+		subpass.pipelineBindPoint    = VK_PIPELINE_BIND_POINT_GRAPHICS;
+		subpass.pColorAttachments    = color_references.data();
+		subpass.colorAttachmentCount = 1;
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: `1` -> `static_cast<uint32_t>(color_references.size())`

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 29]
```diff
@@ -0,0 +1,988 @@
...
+{
+	// Initialize time domain count:
+	uint32_t time_domain_count = 0;
+	// Update time domain count:
+	VkResult result = vkGetPhysicalDeviceCalibrateableTimeDomainsEXT(get_device().get_gpu().get_handle(), &time_domain_count, nullptr);
+
+	if (result == VK_SUCCESS)
+	{
+		// Initialize time domain vector
+		time_domains.clear();
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: No need to clear `time_domains`, as this is called just once.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 30]
```diff
@@ -0,0 +1,988 @@
...
+		// Resize time domains vector:
+		time_domains.resize(static_cast<int>(time_domain_count));        // this needs static cast to int
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Why should you need to cast `time_domain_count` to `int`?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 31]
```diff
@@ -0,0 +1,988 @@
...
+			timestamp_info.timeDomain = time_domain;
+
+			// Push-back timestamp info to timestamps info vector:
+			timestamps_info.push_back(timestamp_info);
+		}
+
+		// Resize time stamps vector
+		timestamps.resize(static_cast<int>(time_domain_count));
+		// Resize max deviations vector
+		max_deviations.resize(static_cast<int>(time_domain_count));
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: No need to `static_cast<int>` here

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 32]
```diff
@@ -0,0 +1,988 @@
...
+	}
+
+	// Time domain is successfully updated:
+	is_time_domain_updated = ((result == VK_SUCCESS) && (time_domain_count > 0));
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: `is_time_domain_updated` sounds like if it is, well, updated, regularly, after some amount of time.
But it is just that everything is initialized/prepared once to actually get the data.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 33]
```diff
@@ -0,0 +1,988 @@
...
+void CalibratedTimestamps::get_timestamps()
+{
+	// Ensures that time domain exists
+	if (is_time_domain_updated)
+	{
+		// Get calibrated timestamps:
+		VkResult result = vkGetCalibratedTimestampsEXT(get_device().get_handle(), time_domains.size(), timestamps_info.data(), timestamps.data(), max_deviations.data());
+
+		// Timestamps is successfully updated:
+		is_timestamp_updated = (result == VK_SUCCESS);
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Maybe change this function to return a bool `(result == VK_SUCCESS)`.
No need to have that `is_timestamp_updated` anymore.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 34]
```diff
@@ -0,0 +1,988 @@
...
+	}
+}
+
+void CalibratedTimestamps::init_time_domains_and_timestamps()
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: Is this function useful at all?
It's just some one-time member variable initialization.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 35]
```diff
@@ -0,0 +1,988 @@
...
+
+	// Now to get timestamps
+	get_timestamps();
+
+	if (is_timestamp_updated)
+	{
+		// Create a local delta_timestamp to push back to the vector delta_timestamps
+		DeltaTimestamp delta_timestamp{};
+
+		// Naming the tic-toc tags and begin timestamp for this particular mark
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: tic-toc ?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 36]
```diff
@@ -0,0 +1,988 @@
...
+		if (!input_tag.empty())
+		{
+			delta_timestamp.tag = input_tag;
+		}
+		delta_timestamp.begin = timestamps[selected_time_domain.index];
+
+		// Push back this partially filled element to the vector, which will be filled in its corresponding toc
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: toc ?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+	} selected_time_domain{};
+
+	struct DeltaTimestamp
+	{
+		uint64_t    begin = 0;
+		uint64_t    end   = 0;
+		uint64_t    delta = 0;
+		std::string tag   = "Untagged";
+
+		void get_delta()
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h]: Is this really worth a function?
If at all, it doesn't get you anything (returning void) but calculates something. So `calculate_delta` or `determine_delta` would be a better fit.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.h [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+class CalibratedTimestamps : public ApiVulkanSample
+{
+  private:
+	bool                         is_time_domain_updated = false;        // this is just to tell if time domain update has a VK_SUCCESS in the end
+	bool                         is_timestamp_updated   = false;        // this is just to tell the GUI whether timestamps operation has a VK_SUCCESS in the end
+	std::vector<VkTimeDomainEXT> time_domains{};                        // this holds all time domains extracted from the current Instance
+	std::vector<uint64_t>        timestamps{};                          // timestamps vector
+	std::vector<uint64_t>        max_deviations{};                      // max deviations vector
+
+	struct
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.h:36]: Some structs are named, others are not.
Is there a rationale behind that?

**Review by asuessenbach — COMMENTED** [review]
This sample just shows how to use the CalibratedTimestamps functions. If that was your goal, I'm fine with it.
But I see three possible improvements:
- With the `std::vector<DeltaTimestamp> delta_timestamps`, you are prepared to take multiple timestamps per frame. But you don't do it. Maybe you should consider adding one or more additional timestamps?
- You explicitly select the device domain, but measure the recording (or building) of a command buffer, which probably happens completely on the host. Maybe you should measure some device work, instead, like drawing or buffer uploading?
- You don't show any use of the deviations. Maybe at least displaying it could get you some feeling on what they are?

**Comment by gpx1000** [comment]
I tend to agree this sample doesn't fully exercise calibrated timestamps.  However, it does a fairly good job of demonstrating how to use them which I think is a decent goal.  I'll have Jeremy do another pass at increasing the scope a little bit more.  My only concern is about ensuring this sample's core is to teach specifically how to do calibrated timestamps and not get too complex to ultimately be useful.  Thanks for the review!

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp (hunk 28 above)
Comment (asuessenbach) [reply samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: You had marked this as resolved without any change. Do you want to keep it like this? If so, why?

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp (hunk 30 above)
Comment (asuessenbach) [reply samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: You should as well remove the comment.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp (hunk 32 above)
Comment (asuessenbach) [reply samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: You should as well adjust the comment.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 39]
```diff
@@ -373,7 +373,7 @@ void CalibratedTimestamps::prepare_offscreen_buffer()
 		render_pass_create_info.sType                  = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
 		render_pass_create_info.pAttachments           = attachment_descriptions.data();
 		render_pass_create_info.attachmentCount        = static_cast<uint32_t>(attachment_descriptions.size());
-		render_pass_create_info.subpassCount           = 1;
+		render_pass_create_info.subpassCount           = static_cast<uint32_t>(color_references.size());
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: the subpassCount is not related to the number of color_references! Keep the `1` here.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp [hunk 40]
```diff
@@ -109,24 +110,24 @@ void CalibratedTimestamps::build_command_buffers()
 	render_pass_begin_info_0.clearValueCount       = static_cast<uint32_t>(clear_values_0.size());
 	render_pass_begin_info_0.pClearValues          = clear_values_0.data();
```
Comment (asuessenbach) [inline samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: One could go even further here, and move all the `VkClearValue`s, `VkRenderPassBeginInfo`s, `VkViewport`s, and `VkRect2D`s out of the loop, renaming them for example to `offscreen*`, `bloom*` or `filter_pass*`, and `framebuffer*`.
Would make the actual recording loop very short and clear.

File: samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp (hunk 28 above)
Comment (gpx1000) [reply samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp]: that's my fault; I could have sworn I got it but I definitely missed it.

**Review by SaschaWillems — DISMISSED** [review]

**Review by asuessenbach — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]
I just added calibrated timestamp support to my driver, so was actually able to test this.

**Review by SaschaWillems — APPROVED** [review]

**Comment by marty-johnson59** [comment]
Merging w/ 3 approvals

**Pull request merged**